### DISCUSSION TITLE
[@types/react-dom] upgrading from 18.2.14 to 18.2.15 "breaks" types useFormState hook

### DISCUSSION THREAD

**Discussion by ethernal** [DefinitelyTyped/DefinitelyTyped#68169] [Issues with a @types package] [answered]
I am certain it's something I do or don't do in my code but when using `useFormState` hook starting in version `18.2.15` (in next 14.0.4) I am getting an error about the type I use in the hook:

```tsx
type ContactFormData = {
	email: string;
	name: string;
	message: string;
};

const initialFormState: ContactFormData = {
		email: '',
		name: '',
		message: '',
	};

	const toastId = React.useRef<Id | null>(null);
	const formRef = React.useRef<HTMLFormElement | null>(null);

	const [state, formAction] = useFormState(
		async (prev: any, formData: FormData) => {
			const res = await sendEmailAction(prev, formData);
			if (res === true) {
				updateMailSuccess();
				formRef.current?.reset();
			} else updateMailError();
		},
		initialFormState,
	);
```

In 18.2.14 there are no issues, but in 18.2.15 things get wild with expecting void for both `action` function and `initialData` that last part makes no sense to me as having void data is no data..

Since this has been quite a while and I found no issues open or closed related to that hook I assume it's something I do that triggers the error but I am not sure what that is. The same code runs in prod just fine, but vercel will not build if linting errors are present.

I could use a hand in figuring out if this is error in types definition (I am not entirely certain) or if this is issue with my code.

To see the difference check the repo for my site: https://github.com/ethernal/personal-site
if you execute `npm install @types/react-dom@18.2.15` code in `ContactForm.tsx` in `src\components\ContactForm\ContactForm.tsx` will get squiggly and build will fail. Going back to 18.2.14 - no issue is found.

Error in 18.2.15:

```
No overload matches this call.
  Overload 1 of 2, '(action: (state: void) => void | Promise<void>, initialState: void, permalink?: string | undefined): [state: void, dispatch: () => void]', gave the following error.
    Argument of type '(prev: any, formData: FormData) => Promise<void>' is not assignable to parameter of type '(state: void) => void | Promise<void>'.
      Target signature provides too few arguments. Expected 2 or more, but got 1.
  Overload 2 of 2, '(action: (state: void, payload: FormData) => void | Promise<void>, initialState: void, permalink?: string | undefined): [state: void, dispatch: (payload: FormData) => void]', gave the following error.
    Argument of type 'ContactFormData' is not assignable to parameter of type 'void'.ts(2769)
```

Thank you.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
This seems to be correct. You forget to return the new state from your action function.

**Comment by ethernal** [comment]
🤦‍♂️ returning state fixes the issue, how did I not see that... Thank you!
